### ISSUE TITLE
Alignment greater than page size locals can cause invalid access exceptions.

### ISSUE THREAD

**Issue by retep998** [rust-lang/rust#70143]
Forked from https://github.com/rust-lang/rust/issues/70022

Minimal example
```Rust
#[repr(align(0x100000))]
struct Aligned(u8);

fn main() {
    let x = Aligned(0);
    println!("{:#x}", &x as *const _ as usize);
}
```
Aligning the stack is done *after* the stack probe. Because stacks grow downwards and aligning the stack shifts it downwards, it can cause the end of the stack to extend past the guard page and cause invalid access exceptions *or worse* when those sections of the stack are touched.

Only confirmed that this occurs on `pc-windows-msvc`.

**Comment by Amanieu** [comment]
This is most likely an LLVM bug.

**Comment by RalfJung** [comment]
Could someone who knows the technical details here report this bug upstream with LLVM?

**Comment by LiamTheProgrammer** [comment]
I think this is platform-specific but it still all happens to varying degrees on x86(-64). This safe code causes a stack overflow on the Rust Playground:

```rust
#[repr(align(0x800000))]
struct Aligned(usize);

fn main() {
    let x = Aligned(2);
    println!("{}", x.0);
}
```

This safe code causes a segfault on the Rust Playground:

```rust
#[repr(align(0x10000000))]
struct Aligned(usize);

fn main() {
    let x = Aligned(2);
    println!("{}", x.0);
}
```

I don't know how this would be done, but maybe doing such alignments should only be allowed if `unsafe` is used to keep it out of safe code? Embedded might depend on high alignments, so I don't think removing the ability to align with high numbers entirely would be a good solution.

**Comment by RalfJung** [comment]
Note that we do install a stack guard and associated handler to check for stack overflows. So just getting a stack overflow or segfault does not necessarily demonstrate a problem.

**Comment by retep998** [comment]
> The bug here is about the locals not having the alignment that was requested via `repr`.

The locals do absolutely have the required alignment. The issue is with the stack aligning prologue which can cause the function's frame to exist entirely past the guard page so unrelated memory can be stomped on.

**Comment by cuviper** [comment]
I just tried this on godbolt: https://rust.godbolt.org/z/8xfn4v

```rust
#[repr(align(0x10000000))]
struct Aligned(usize);

pub fn foo() -> usize {
    Aligned(2).0
}
```

The output for `x86_64-pc-windows-msvc`:

```asm
example::foo:
        push    rbp
        mov     eax, 536870896    # 0x1ffffff0
        call    __chkstk
        sub     rsp, rax
        lea     rbp, [rsp + 128]
        and     rsp, -268435456    # 0xfffffffff0000000
        mov     qword ptr [rsp], 2
        mov     rax, qword ptr [rsp]
        lea     rsp, [rbp + 536870768]    # 0x1fffff70
        pop     rbp
        ret
```

In the worst case, the `sub` could move `rsp` to the very top of an aligned block, or anywhere more than a page size away from the aligned base. Then the large `and` will round this down, which might jump over the stack guard page, and this wasn't tested by `__chkstk`.

The output for `x86_64-unknown-linux-gnu`:

```asm
example::foo:
        push    rbp
        mov     rbp, rsp
        and     rsp, -268435456    # 0xfffffffff0000000
        mov     eax, 536870912    # 0x20000000
        call    __rust_probestack
        sub     rsp, rax
        mov     qword ptr [rsp], 2
        mov     rax, qword ptr [rsp]
        mov     rsp, rbp
        pop     rbp
        ret
```

Here the alignment is done first, but it's still bad. We might _start_ with `rsp` more than a page size unaligned, and then again the large `and` can jump over the guard page. Then `__rust_probestack` could be probing non-stack memory and think everything is fine.

Here's one more Linux test I ran locally, using LLVM 11's `"probe-stack"="inline-asm"`:

```asm
example::foo:
        pushq   %rbp
        movq    %rsp, %rbp
        andq    $-268435456, %rsp    # 0xfffffffff0000000
        movq    %rsp, %r11
        subq    $536870912, %r11    # 0x20000000
.LBB0_1:
        subq    $4096, %rsp
        movq    $0, (%rsp)
        cmpq    %r11, %rsp
        jne     .LBB0_1
        movq    $2, (%rsp)
        movq    (%rsp), %rax
        movq    %rbp, %rsp
        popq    %rbp
        retq
```

Again, the initial alignment could jump the guard page, then the probe loop may be clobbering non-stack memory.
(cc @serge-sans-paille)

**Comment by Amanieu** [comment]
It seems that the ideal behavior here would be to add the alignment to the size passed to `__chkstk`/`__rust_probestack` and perform the stack alignment masking after the stack probing.

**Comment by cuviper** [comment]
Yeah, I've been talking privately with Serge, and he's going to try implementing something like that.

**Comment by RalfJung** [comment]
I thought this is what already happens?

> Aligning the stack is done after the stack probe. 

And the fix would be to align first and then run the stack probe?

**Comment by retep998** [comment]
> 
> 
> I thought this is what already happens?
> 
> > Aligning the stack is done after the stack probe.
> 
> And the fix would be to align first and then run the stack probe?

The fix is to *first* probe the stack for the frame size *plus the alignment*, and then align the stack. Currently stack probes only check the frame size without adding the alignment to that size.

**Comment by cuviper** [comment]
Note that the size of the alignment adjustment is a dynamic value -- the offset from whatever incoming stack pointer we get to an aligned pointer below that. Then the actual frame size is allocated after we have known alignment. If the stack guard page is anywhere in that total range, we need to fault, which is what the probing does.

**Comment by RalfJung** [comment]
I guess I am wondering why we can't first align and then probe for the already computed new stack size, that sounds like it avoids doing the aligning twice. But it probably doesn't work for other reasons I cannot see.

**Comment by retep998** [comment]
@RalfJung Because after aligning the current stack pointer can be *past* the guard page in a completely unrelated section of memory, that could be completely valid memory for the full size of the stack frame that the stack probe could succeed on, but it's not our stack and we'd be trampling that unrelated memory!

**Comment by RalfJung** [comment]
Ah I see, I somehow assumed stack probing would still have access to the old top-of-stack and could thus probe all the way from there to the aligned stack with the new frame.

**Comment by cuviper** [comment]
The Linux examples I gave do have the old pointer in `rbp`. However, once we've moved `rsp`, a signal could be delivered and run its handler from that stack location, before we've probed that it's safe.

**Comment by cuviper** [comment]
I think we can now consider this a Windows-only bug, because the rest are now using `"probe-stack"="inline-asm"`, except for arches that lack stack probes entirely. It's not really specific to MSVC either, as `windows-gnu` generates the same code calling `__chkstk`.

@rustbot label -O-linux -O-windows-msvc +O-windows

**Comment by apiraino** [comment]
Visited during [T-compiler P-high review](https://rust-lang.zulipchat.com/#narrow/channel/238009-t-compiler.2Fmeetings/topic/.5Btriage.5D.202026-06-19.20P-high.20review/near/604900802)[#t-compiler/meetings > &#91;triage&#93; 2026-06-19 P-high review @ 💬](https://rust-lang.zulipchat.com/#narrow/channel/238009-t-compiler.2Fmeetings/topic/.5Btriage.5D.202026-06-19.20P-high.20review/near/604900802)

Looks nothing changed here and the fix is on us rather than LLVM? cc: @rust-lang/wg-llvm

**Comment by ChrisDenton** [comment]
[recent godbolt](https://rust.godbolt.org/#g:!((g:!((g:!((h:codeEditor,i:(filename:'1',fontScale:14,fontUsePx:'0',j:1,lang:rust,selection:(endColumn:44,endLineNumber:7,positionColumn:44,positionLineNumber:7,selectionStartColumn:44,selectionStartLineNumber:7,startColumn:44,startLineNumber:7),source:'%23%5Brepr(align(0x10000))%5D%0Astruct+Aligned(u8)%3B%0A%0A%23%5Binline(never)%5D%0Apub+fn+over_aligned()+%7B%0A++++let+x+%3D+Aligned(0)%3B%0A++++println!!(%22%7B:%23x%7D%22,+%26x+as+*const+_+as+usize)%3B%0A%7D'),l:'5',n:'0',o:'Rust+source+%231',t:'0')),k:50,l:'4',n:'0',o:'',s:0,t:'0'),(g:!((g:!((h:compiler,i:(compiler:r1960,filters:(b:'0',binary:'1',binaryObject:'1',commentOnly:'0',debugCalls:'1',demangle:'0',directives:'0',execute:'1',intel:'0',libraryCode:'0',trim:'1',verboseDemangling:'0'),flagsViewOpen:'1',fontScale:14,fontUsePx:'0',j:1,lang:rust,libs:!(),options:'--target+x86_64-pc-windows-msvc',overrides:!((name:edition,value:'2024')),selection:(endColumn:1,endLineNumber:1,positionColumn:1,positionLineNumber:1,selectionStartColumn:1,selectionStartLineNumber:1,startColumn:1,startLineNumber:1),source:1),l:'5',n:'0',o:'+rustc+1.96.0+(Editor+%231)',t:'0')),k:50,l:'4',m:50,n:'0',o:'',s:0,t:'0'),(g:!((h:ir,i:('-fno-discard-value-names':'0',compilerName:'rustc+1.96.0',demangle-symbols:'0',editorid:1,filter-attributes:'0',filter-comments:'0',filter-debug-info:'0',filter-declarations:'1',filter-instruction-metadata:'0',filter-library-functions:'1',fontScale:14,fontUsePx:'0',j:1,selection:(endColumn:1,endLineNumber:1,positionColumn:1,positionLineNumber:1,selectionStartColumn:1,selectionStartLineNumber:1,startColumn:1,startLineNumber:1),show-optimized:'0',treeid:0,wrap:'1'),l:'5',n:'0',o:'LLVM+IR+Viewer+rustc+1.96.0+(Editor+%231,+Compiler+%231)',t:'0')),header:(),l:'4',m:50,n:'0',o:'',s:0,t:'0')),k:50,l:'3',n:'0',o:'',t:'0')),l:'2',n:'0',o:'',t:'0')),version:4)

**Comment by RalfJung** [comment]
This is still an LLVM bug, isn't it? We just gave up on hoping they'll fix it so we plan to reject it on the Rust side?

**Comment by ChrisDenton** [comment]
Yeah. Making it an error unless it is ever fixed in llvm was the plan. Notably both `clang` and [msvc's `cl`](https://learn.microsoft.com/en-us/cpp/error-messages/compiler-errors-1/compiler-error-c2345?view=msvc-170) have a limit of 8192 alignment for the target ([godbolt](https://rust.godbolt.org/#g:!((g:!((g:!((h:codeEditor,i:(filename:'1',fontScale:14,fontUsePx:'0',j:1,lang:c%2B%2B,selection:(endColumn:30,endLineNumber:1,positionColumn:30,positionLineNumber:1,selectionStartColumn:30,selectionStartLineNumber:1,startColumn:30,startLineNumber:1),source:'struct+alignas(0x10000)+S+%7B%7D%3B'),l:'5',n:'0',o:'C%2B%2B+source+%231',t:'0')),k:50,l:'4',n:'0',o:'',s:0,t:'0'),(g:!((g:!((h:compiler,i:(compiler:clang_trunk,filters:(b:'0',binary:'1',binaryObject:'1',commentOnly:'0',debugCalls:'1',demangle:'0',directives:'0',execute:'1',intel:'0',libraryCode:'0',trim:'1',verboseDemangling:'0'),flagsViewOpen:'1',fontScale:14,fontUsePx:'0',j:1,lang:c%2B%2B,libs:!(),options:'-target+x86_64-pc-windows-msvc',overrides:!(),selection:(endColumn:1,endLineNumber:2,positionColumn:1,positionLineNumber:2,selectionStartColumn:1,selectionStartLineNumber:2,startColumn:1,startLineNumber:2),source:1),l:'5',n:'0',o:'+x86-64+clang+(trunk)+(Editor+%231)',t:'0')),header:(),k:50,l:'4',m:50,n:'0',o:'',s:0,t:'0'),(g:!((h:output,i:(compilerName:'x86-64+gcc+16.1',editorid:1,fontScale:14,fontUsePx:'0',j:1,wrap:'1'),l:'5',n:'0',o:'Output+of+x86-64+clang+(trunk)+(Compiler+%231)',t:'0')),header:(),l:'4',m:50,n:'0',o:'',s:0,t:'0')),k:50,l:'3',n:'0',o:'',t:'0')),l:'2',n:'0',o:'',t:'0')),version:4)). In my (brief) testing, 16384 appeared to work in rust though it's possible I'm missing some situation where that would break.
